### ISSUE TITLE
[macOS] Flow was not found when attempting to start

### ISSUE THREAD

**Issue by bibekg** [facebook/nuclide#1271]
### Issue and Steps to Reproduce
Despite having installed Flow both globally and as a dev dependency, and despite Flow working as expected in the Terminal when using both the commands `yarn run flow` and `flow`, every time I open Atom/Nuclide, I get the warning saying "Flow was not found when attempting to start it in [this directory]". I've also checked the box in Nuclide settings for `Use the Flow binary provided in each project's flow-bin`

A few more things I noticed from digging around, although I'm not sure how relevant they are:

Looking in the Nuclide Health stats, Nuclide seems to be running `which <PATH_TO_FLOW>` a lot. Not sure how relevant it is though. It's also running `which hh_client`, after which it doesn't `which` anything else.
![image](https://user-images.githubusercontent.com/8083680/29497891-90d888f4-85a5-11e7-9aab-3b7e3cf375ee.png)

Running the same command in Terminal returns the same directory, e.g. running `which <DIR>` echoes `<DIR>`

### Expected Behavior

Flow should start in Nuclide and start showing errors in-line.

### Actual Behavior

Flow does not start and shows the warning that it could not be found.

### Versions
* Atom: 1.19.2
* Nuclide: 0.245.0
* Client OS: macOS 10.12.6 (Sierra)
* Server OS (optional): N/A

### Additional Details
* Installed packages (`apm ls --installed`):

![image](https://user-images.githubusercontent.com/8083680/29497877-44bdab84-85a5-11e7-9a21-d74dc9d6124a.png)

**Comment by matthewwithanm** [comment]
Hi @bibekg! Have you tried setting the `nuclide.pathToFlow` ("Path to Flow Executable") setting?

**Comment by bibekg** [comment]
@matthewwithanm Hey, thanks for the response. I haven't tried that yet. I'll give that a shot. Ideally though, I don't want to set the path explicitly. I'd prefer to use the flow-bin provided in each project if possible.

**Comment by matthewwithanm** [comment]
Yeah, that makes sense. I'll ask @nmote to look at it next week when he gets back 😊

**Comment by bibekg** [comment]
Another interesting thing. I tested on a different macOS machine running the same version of Atom and Nuclide and it shows inline Flow errors for one project but not for another. This is with completely empty `.flowconfig` files in both projects.

**Comment by bibekg** [comment]
I think I figured it out. I did some experimenting and ran `brew install flow` and `brew link --overwrite flow` and it's now working. I never saw any mention of needing to do this on the FlowType website/installation instructions. Is this an expected requirement?

**Comment by go1t** [comment]
I'm using remote development and got the same issue. There was no issue when it was v0.244, so I assume it's something related to 0.245 or 0.246

**Comment by xdissent** [comment]
I was facing this issue, but realized it had nothing to do with nuclide in my case. I use nodenv, and had globally installed `flow-upgrade` in a particular nodenv version. That seems to install a bin for the `which` pkg as well. With nodenv global bins, a shim is always installed in the PATH, regardless of whether the currently-active nodenv version contains the package. Running it from an incorrect nodenv version will just error with a message telling you which versions contain the actual pkg. On macOS, the native `which` is in `/usr/bin`, which is *after* the nodenv shims in my PATH. So nuclide was using the shim, but from a nodenv version without the `which` pkg. Potential fixes include adjusting the PATH, installing `which` in the correct nodenv version, or uninstalling `flow-upgrade` from all nodenv versions.

TL;DR - make sure nuclide is using the correct `which` bin.

**Comment by petrjaros** [comment]
Uninstalling `flow-upgrade` worked for me. Thanks @xdissent!

**Comment by petrjaros** [comment]
After I upgraded Nuclide to 0.249.0 it has stopped working again. So to have flow working in Atom again I had to downgrade to 0.246.0 and uninstall `flow-upgrade`. BTW I use Windows so the issue is not only Mac related.

**Comment by Tsury** [comment]
How did you downgrade?

**Comment by Fity** [comment]
doesn't work for nuclide@0.249.0 and atom@1.19.4 on Mac OSX10.12.6.
`flow-bin` is in my devRequirements(works fine with yarn run flow).
I didn't install `flow-upgrade` and both install flow globally(with `brew install flow`) and specify the flow bin path can't make it work.(Of course `use the flow binary included in each project's flow-bin` option is checked)

any help is welcome.

some extra info:
```
Community Packages (21) /Users/daofengyang/.atom/packages
├── MagicPython@1.0.12
├── editorconfig@2.2.2
├── file-icons@2.1.11
├── haskell-grammar@0.4.0
├── highlight-selected@0.13.1
├── language-babel@2.73.0
├── language-graphql@0.9.0
├── language-ini@1.19.0
├── language-lua@0.9.11
├── language-ocaml@1.9.5
├── language-rust@0.4.12
├── language-swift@0.5.0
├── language-thrift@1.0.2
├── linter@2.0.0
├── linter-eslint@8.2.1
├── nuclide@0.249.0
├── nuclide-format-js@0.0.41
├── prettier-atom@0.38.0
├── set-syntax@0.3.2
├── sort-lines@0.15.0
└── tool-bar@1.1.0
```

`"flow-bin": "^0.47.0",` is in my `package.json` and `.flowconfig` exists in my project's root directory.

**Comment by bibekg** [comment]
I also don't have `flow-upgrade` installed.

**Comment by nmote** [comment]
I can't repro on my OS X laptop, using the "Use the Flow binary included in each project's flow-bin" setting with Nuclide on master as well as Flow v0.53.1 installed in the project by `yarn`. Flow starts up and runs just fine for me.

I also don't see any changes related to how Nuclide finds or executes Flow between v0.244.0 (which several people here have reported as working) and v0.249.0 (https://github.com/facebook/nuclide/compare/release-v0.244.0...release-v0.249.0)

This leads me to believe that this issue is caused by something external to Nuclide and Flow. If folks have a different binary for `which` like @xdissent did, I could definitely see that causing problems. I'd see what the output of `which which` (lol) is -- if it's not `/usr/bin/which` (on OS X, anyway) then something weird is definitely going on.

You may also be able to shed some light on the situation by posting the logs after this happens. I believe they are located in `$TMPDIR/nuclide-$USER-logs`. Please put them in a gist or use a paste service, and include the link here. Don't paste them directly into a comment.

**Comment by bibekg** [comment]
My `which which` result isn't exactly what you said but it doesn't look like it should be an issue.
```
> which which
/usr/local/bin/which
```

Here's a [gist](https://gist.github.com/bibekg/024de971058407fa8995d4395ff12cb4) of my `nuclide.log` file.

**Comment by petrjaros** [comment]
I tried it again. Nuclide 0.250.0 can't find Flow (I tried Flow 0.52.0 and 0.53.1). My Flow is installed globally by npm and locally in package.json. "Use the Flow binary included in each project's flow-bin" setting does not make any difference. My `nuclide.log` is nearly identical to the one @bibekg posted. I use Windows so no `which` is preset in the system. When I downgrade Nuclide to 0.246.0 Flow works as usual.

**Comment by nmote** [comment]
I suspect the Windows problems are unrelated. Windows support has been a bit dicey for a while, and #1041 changed things in v0.249.0. It seems possible that it breaks some cases (cc @ebluestein @jrajav). Check out #1281 -- it looks like there's a fix for that coming in #1280.

**Comment by ebluestein** [comment]
Yeah #1041 regressed locating flow for certain configurations, a fix is on the way.

**Comment by Fity** [comment]
A interesting fix for my case is, open atom within my terminal(with atom cli).
By the way, I am using the fish shell, and atom cli shell uses bash/sh, any ideas about this?

**Comment by bibekg** [comment]
Reinstalled Atom using the following steps and it works now:
1. delete the app from the Applications folder.
2. delete the invisible .atom directory from the home folder.
3. delete the /usr/local/bin/atom tool.
4. delete the /usr/local/bin/apm tool.
5. delete the ~/Library/Preferences/com.github.atom.plist file.
6. install Atom from [atom.io](https://atom.io).
7. install Nuclide via the packages pane in atom

Not sure what changed as a result of this but I'll take it. 😄

**Comment by kunokdev** [comment]
I had similar issue on both Linux and Mac. Both times searching for solution it led me to this issue. So if someone gets similar problem; after fresh install of Nuclide, even tho `flow` works within terminal, but IDE can't find it, here is solution that solved it for me on both platforms.

First run this command to get path to Flow:
```
which flow
```
it should output something like this:
```
/usr/local/bin/flow
```
then on Ubuntu you can run `vim ~/.bashrc` or else on Mac you can run `vim ~/.bash_profile` to edit file in which you can extend PATH variable. If you don't like Vim, use anything else that can edit text.

Then add this line:
```
export PATH='$PATH:/usr/local/bin/flow'
```
Restart session and run `echo $PATH` and check if added line is included.

Alternatively, on Mac you can run `sudo vim /etc/paths` instead and add new line to that file with output from `which flow` .
